torch/csrc/api/include/torch/nn/modules/any.h

```diff
@@ -0,0 +1,215 @@
+/// invoked, with dynamic type checking.
+class AnyModule {
+ public:
+  /// Constructs an `AnyModule` from a `shared_ptr` to concrete module object.
+  template <typename ModuleType>
+  explicit AnyModule(std::shared_ptr<ModuleType> module)
+      : content_(make_holder(
+            std::move(module),
+            &std::remove_reference<ModuleType>::type::forward)) {}
+
+  /// Constructs an `AnyModule` from a concrete module object.
+  template <typename ModuleType>
+  explicit AnyModule(ModuleType&& module)
+      : AnyModule(
+            std::make_shared<ModuleType>(std::forward<ModuleType>(module))) {}
+
+  /// Move construction and assignment is allowed, and follows the default
+  /// behavior of move for `std::unique_ptr`.
+  AnyModule(AnyModule&&) = default;
+  AnyModule& operator=(AnyModule&&) = default;
+
+  /// Copy is disallowed.
+  AnyModule(const AnyModule& other) = delete;
+  AnyModule& operator=(const AnyModule& other) = delete;
+
+  /// Invokes `forward()` on the contained module with the given arguments, and
+  /// returns the return value as an `Any`. Use this method when chaining
+  /// `AnyModule`s in a loop.
+  template <
+      typename... ArgumentTypes,
+      typename = disable_for_t<Any, ArgumentTypes...>>
+  Any forward(ArgumentTypes&&... arguments) {
+    return content_->forward(std::forward<ArgumentTypes>(arguments)...);
+  }
+
+  /// A type erased `forward()` that invokes the actual `forward()` with a value
+  /// of any type, and returns a type erased object.
+  Any forward(Any&& any) {
+    std::vector<Any> vector;
+    vector.push_back(std::move(any));
+    return content_->forward(std::move(vector));
+  }
+
+  /// Attempts to cast the underlying module to the given module type. Throws an
+  /// exception if the types do not match.
+  template <typename T, typename = enable_if_module<T>>
+  T& get() {
+    return get_<T>();
+  }
+
+  /// Attempts to cast the underlying module to the given module type. Throws an
+  /// exception if the types do not match.
+  template <typename T, typename = enable_if_module<T>>
+  const T& get() const {
+    return get_<T>();
+  }
+
+  /// Returns a `std::shared_ptr` whose dynamic type is that of the underlying
+  /// module.
+  std::shared_ptr<Module> ptr() const {
+    return content_->ptr();
+  }
+
+  /// Like `ptr()`, but casts the pointer to the given type.
+  template <typename T, typename = enable_if_module<T>>
+  std::shared_ptr<T> ptr() const {
+    /// Call get() but discard the value, just to do the type checking.
+    get<T>();
+    return std::static_pointer_cast<T>(ptr());
+  }
+
+  /// Returns the `type_info` object of the contained value.
+  const std::type_info& type_info() const noexcept {
+    return content_->type_info;
+  }
+
+ private:
+  /// The static type of the object we store in the `AnyModule`, which erases
+  /// the actual type, but allows us to call `forward()` on the underlying
+  /// module.
+  struct Placeholder {
```

Comment:
I wouldn't implement this stuff twice. Have you considered using an `Any` member inside of `AnyModule` to solve this?

Reply:
I can see if refactoring out `Any`'s `Placeholder` makes the code shorter. Assuming we stick with our own `Any`